source/utils/flag_manager.hpp

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include "bitset.hpp"
+#include "packet.hpp"
+
+#include <unordered_map>
+#include <spdlog/fmt/fmt.h>
+
+#include "logger/logger.hpp"
+
+namespace sim {
+
+template <typename FlagId, BitStorageType BitStorage>
+class FlagManager {
+public:
+    FlagManager() : m_next_pos(0) {}
+
+    bool register_flag_by_amount(FlagId id, BitStorage different_values) {
+        return register_flag_by_length(id, required_bits_for_values(different_values));
+    }
+
+    bool register_flag_by_length(FlagId id, BitStorage flag_length) {
+        if (flag_length == 0 || flag_length > sizeof_bits(BitStorage)) {
+            LOG_ERROR(fmt::format("Incorrect flag length. Max possible length is {}, flag length should be more than 0. Got {}", sizeof_bits(BitStorage), flag_length));
+            return false;
+        }
```

Comment:
Second condition here is redundant due to more stricter check in next `if`:

```suggestion
        if (flag_length == 0) {
            LOG_ERROR("Passed zero flag length, should be at least 1");
            return false;
        }
```